feat(drift_dev): Support view triggers in drift SQL

dnys1 opened the pull request:
Adds support for triggers on views created by `CREATE TRIGGER` expressions.

https://www.sqlite.org/lang_createtrigger.html#instead_of_triggers

dnys1 committed:
feat: Support view triggers in drift SQL

Adds support for triggers on views created by `CREATE TRIGGER` expressions.

https://www.sqlite.org/lang_createtrigger.html#instead_of_triggers

simolus3 reviewed (commented):
Thanks! I think this will need some minor additions before it fully works though: In most builds, drift serializes analysis results as intermediate files to speed things up. There's a cast in `ElementDeserializer._readDriftElement` that also affects triggers in line 632.

In addition to the `TestBackend`-based tests which only exercise drift's analyzer, a simple e2e should give confidence that nothing breaks in later steps. Something like this would work with a simple assertion checking that the generated trigger uses the right view getter in generated code:

https://github.com/simolus3/drift/blob/9d41c821eeaa28054f13c7e706b4feb669f98374/drift_dev/test/backends/build/build_integration_test.dart#L404-L430

dnys1 committed:
Fix update kind for empty updates

Iterable.every always returns `true` when the iterable is empty. From what I can tell, it is better to return `false` in these cases.

dnys1 committed:
Fix deserialization of IR for triggers on views

dnys1 commented on drift_dev/lib/src/analysis/resolver/resolver.dart:
```diff
@@ -48,7 +48,7 @@ class DriftResolver {
       }
     } on CouldNotDeserializeException catch (e, s) {
-      driver.backend.log.fine('Could not deserialize $element', e, s);
+      driver.backend.log.warning('Could not deserialize $element', e, s);
```
In order to have the test fail for deserialization exceptions, I needed to change this to a warning.

simolus3 replied:
Which is appropriate, considering that this shouldn't fail :+1:

dnys1 commented on drift_dev/lib/src/analysis/results/query.dart:
```diff
@@ -398,9 +398,11 @@ class UpdatingQuery extends SqlQuery {
   final AstNode root;
 
-  bool get isOnlyDelete => updates.every((w) => w.kind == UpdateKind.delete);
+  bool get isOnlyDelete =>
+      updates.isNotEmpty && updates.every((w) => w.kind == UpdateKind.delete);
```
Since `Iterable.every` returns `true` for empty iterables, this was always true for views leading to generated code with `updateKind: UpdateKind.delete` even for `INSERT` and `UPDATE` statements. My sense was that this was not intentional, but let me know if it was.

simolus3 replied:
I agree that this is kind of weird, but can we move the `isNotEmpty` check to `QueryWriter._writeUpdateKind`?

The intended behavior is that for writes we don't recognize, we don't generate an `updateKind` parameter at all. So I think `_writeUpdateKind` should just check for the list being empty and not emit anything in that case.

I have also improved the situation somewhat in ed72f6fa19d6ef844030c0c87c29f2d107e1483e by including views in `updates`. With this PR landing then, we should be able to recognize these patterns for stream queries as well (an insert on the view in `.drift` files propagates to the table thanks to the trigger, and that then updates everything listening on that).

dnys1 commented:
Good catch! I've added integration tests to verify the output and updated a couple smaller things as noted.

simolus3 reviewed (commented):
Thanks for the tests (and for catching the `isOnlyUpdate`/ `isOnlyDelete` issue). I've added support for recognizing writes to views on `develop` so this should be less of an issue now.

dnys1 committed:
Move check

dnys1 commented:
@simolus3 Makes sense, thanks! Updated 👍

simolus3 reviewed (approved):
Thanks for all your contributions!

dnys1 commented:
My pleasure! Thanks for drift 💪😎
